R/utilities-mass-balance.R

```diff
@@ -357,3 +318,162 @@ getMassBalanceDataMapping <- function(plotType) {
+      molecule$path
+    })
+    # Exclusion criteria
+    excludedMoleculePaths <- NULL
+    # If ExcludePreviousGroupings is TRUE, exclude molecules from previous groupings
+    if(group$ExcludePreviousGroupings %||% TRUE){
+      excludedMoleculePaths <- previouslyIncludedMoleculePaths
+    }
+    if (!is.null(group$Exclude)) {
+      excludedMolecules <- ospsuite::getAllMoleculesMatching(group$Exclude, simulation)
+      excludedMoleculePaths <- c(
+        excludedMoleculePaths,
+        sapply(excludedMolecules, function(molecule){molecule$path})
+      )
+    }
+    includedMolecules <- includedMolecules[!includedMoleculePaths %in% excludedMoleculePaths]
+    includedMoleculePaths <- setdiff(includedMoleculePaths, excludedMoleculePaths)
+    
+    if(isEmpty(includedMoleculePaths)) {
+      next
+    }
+    validateMoleculesFromCompounds(includedMolecules, compoundNames)
+    checkMoleculesAlreadyIncluded(includedMoleculePaths, previouslyIncludedMoleculePaths)
+    
+    previouslyIncludedMoleculePaths <- c(
+      previouslyIncludedMoleculePaths,
+      includedMoleculePaths
+    )
+
+    # For included molecule paths, get their amount vs time and sum them
+    moleculeAmounts <- sapply(
+      includedMoleculePaths,
+      function(moleculePath) {
+        simulationResults$getValuesByPath(moleculePath, individualIds = simulationResults$allIndividualIds)
+      }
+    )
+    moleculeAmounts <- rowSums(as.data.frame(moleculeAmounts))
+    
+    massBalanceData <- rbind.data.frame(
+      massBalanceData,
+      data.frame(
+        Time = simulationResults$timeValues,
+        Amount = moleculeAmounts,
+        Legend = group$Name,
+        stringsAsFactors = FALSE
+      )
+    )
+  }
+  return(massBalanceData)
+}
+
+#' @title defaultMassBalanceGrouping
+#' @description Get mass balance default inclusion/exclusion defined as groupings list
+#' In the Matlab version, the grouping was as followed:
+#' 1) Saliva is excluded from the mass balance calculation
+#' 2) Lumen is a parent compartment that regroups many sub-compartments but Feces
+#' 3) All the rest of the sub-compartments are used normally
+#' @param simulationResults A `SimulationResults` object
+#' @param compoundNames Names of simulation molecules
+#' @import ospsuite
+#' @keywords internal
+defaultMassBalanceGroupings <- function(simulationResults, compoundNames) {
+  groupNames <- unique(sapply(
+    simulationResults$allQuantityPaths,
+    function(path) {
+      head(tail(ospsuite::toPathArray(path), 2), 1)
+    }
+  ))
+
+  defaultGroupings <- c(
+    lapply(
+      groupNames,
+      function(groupName) {
+        list(
+          Name = paste(paste(compoundNames, collapse = ", "), groupName, sep = " - "),
+          Include = paste0("Organism|**|", groupName, "|", compoundNames),
+          # 1) Exclude saliva and 2) Exclude Lumen
+          Exclude = c(
+            paste0("Organism|Saliva|**|", compoundNames),
```

Comment:
Saliva exclusion does not make sense here.
At the moment it's excluded everywhere
![grafik](https://github.com/Open-Systems-Pharmacology/OSPSuite.ReportingEngine/assets/25061876/d013d9fc-31c7-4ef3-8f22-933bfde884ee)

Reply:
Instead, `groupNames `must be reduced.
At the moment we get the following group names for a standard simulation:
![grafik](https://github.com/Open-Systems-Pharmacology/OSPSuite.ReportingEngine/assets/25061876/41fca7cd-4f0b-4030-a97d-c638668c5f23)
Highlighted group names should be removed (Saliva+SalivaryGland+Lumen_segements)

Actually, it's easier to define the group names of interest manually, without asking the model:
* "Plasma"
* "BloodCells"
* "Interstitial"
* "Intracellular"
* "Endosome"
* "Gallbladder"
* "Urine"
* "Feces"

Reply:
Finally: do we have the "Rest" grouping in the default grouping?
If not: it should be added.
And here: exclude Saliva :)